### PR TITLE
Separate schema and form config

### LINKED ISSUE #98: Separate schema and form config

### PR DESCRIPTION
**Pull request by IgorZiegler**
- Add json_reader service for format json before init form
 - Modify json configs

Fixes #98

### CODE REVIEW

File: frontend/blocks/modal/modal.js [hunk 1]
```diff
@@ -35,7 +37,7 @@ export default class Modal extends Base {
   }
 
   get value() {
-    return Object.assign(this._bodyForm.value, this._footerForm.value);
+    return assign(this._bodyForm.value, this._footerForm.value);
```
Comment (102) [inline frontend/blocks/modal/modal.js]: Can we still use `Object.assign` instead?

File: public/examples/forms/text.json [hunk 2]
```diff
@@ -20,7 +21,7 @@
         }
       ]
     },
-    "accountNumber1": {
+    "accounNumber1": {
```
Comment (102) [inline public/examples/forms/text.json]: Typo in `"accountNumber1"`

File: frontend/blocks/container/container.js [hunk 3]
```diff
@@ -25,57 +23,36 @@ export default class Container extends Base {
...
-    mapValues(this.items, block => {
+    this.items.forEach(block => {
       block.render();
       this.appendChild(block);
     });
   }
 
   getItemById(id) {
-    return this.itemsValues.find(item => id === item.id);
+    return this.items.find(item => { return item.id === id; });
```
Comment (102) [inline frontend/blocks/container/container.js]:
```js
return this.items.find(item => item.id === id);
```

File: frontend/blocks/container/container.js [hunk 4]
```diff
@@ -25,57 +23,36 @@ export default class Container extends Base {
...
   }
 
   validate() {
-    return this.itemsValues.every(item => item.validate());
+    return this.items.every(block => {
+      return block.validate();
+    });
```
Comment (102) [inline frontend/blocks/container/container.js]:
```js
return this.items.every(block => block.validate());
```

File: frontend/blocks/container/container.js [hunk 5]
```diff
@@ -25,57 +23,36 @@ export default class Container extends Base {
...
   }
 
   get isValid() {
-    return this.itemsValues.every(item => item.isValid);
+    return this.items.every(block => {
+      return block.isValid;
+    });
```
Comment (102) [inline frontend/blocks/container/container.js]:
```js
return this.items.every(block => block.isValid);
```

File: frontend/blocks/container/container.js [hunk 6]
```diff
@@ -25,57 +23,36 @@ export default class Container extends Base {
...
-  get isPrimitiveValue() {
-    return false;
-  }
-
-  get itemsKeys() {
-    return Object.keys(this.items);
-  }
+    this.items.forEach(block => {
+      Object.assign(result, block.value);
+    });
```
Comment (102) [inline frontend/blocks/container/container.js]:
```js
this.items.forEach(block => Object.assign(result, block.value));

File: frontend/blocks/container/container.js [hunk 7]
```diff
@@ -86,7 +63,9 @@ export default class Container extends Base {
   }
 
   afterRender() {
-    mapValues(this.items, block => block.afterRender());
+    this.items.forEach(block => {
+      block.afterRender();
+    });
```
Comment (102) [inline frontend/blocks/container/container.js]:
```js
this.items.forEach(block => block.afterRender());
```

File: frontend/services/json_reader.js [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+
+  return finalForm;
+}
+
+function configureFormItemsBySchema(items, config) {
+  const fieldsetBlocks = [
+    'address'
+  ];
+
+  return reduce(items, (acc, containerItem) => {
```
Comment (102) [inline frontend/services/json_reader.js]: Do we still can't use `items.reduce` instead?

File: frontend/services/json_reader.js [hunk 9]
```diff
@@ -0,0 +1,64 @@
+import isString from 'lodash/isString';
+import cloneDeep from 'lodash/cloneDeep';
+import map from 'lodash/map';
+
+function formatFormFieldsItemsToArray(form) {
+  const finalForm = cloneDeep(form);
+
+  Object.keys(finalForm).forEach(key => {
+    if (finalForm[key].items && !Array.isArray(finalForm[key].items)) {
```
Comment (mvahrameev) [inline frontend/services/json_reader.js]: I am not sure but I think this may be simplified with `filter`

File: frontend/services/json_reader.js [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+
+  return items.reduce((acc, containerItem) => {
+    if (isString(containerItem)) {
+      if (!config.form[containerItem]) {
+        return acc;
+      }
+
+      acc.push(!fieldsetBlocks.includes(config.form[containerItem].block)
+        ? Object.assign({ name: containerItem }, config.form[containerItem])
+        : config.form[containerItem]);
```
Comment (mvahrameev) [inline frontend/services/json_reader.js]: It will be better to create a variable for `config.form[containerItem]`

File: frontend/index.js [hunk 11]
```diff
@@ -16,7 +16,7 @@ import './blocks/address/address';
 import Modal from './blocks/modal/modal';
 import Form from './blocks/form/form';
 
-import jsonReader from './services/json_reader';
+import jsonReader from './services/config_reader';
```
Comment (102) [inline frontend/index.js]:
```js
import configReader from './services/config_reader';
```

File: frontend/services/config_reader.js [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+      }
+
+      acc.push(!fieldsetBlocks.includes(config.form[containerItem].block)
+        ? Object.assign({ name: containerItem }, config.form[containerItem])
+        : config.form[containerItem]);
+
+      return acc;
+    }
+
+    if (containerItem.items) {
```
Comment (mvahrameev) [inline frontend/services/config_reader.js]: `containerItems = containerItems && ...`

File: frontend/services/config_reader.js [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+    containerItem.items = containerItem.items && configureFormItemsBySchema(containerItem.items, config);
+
+    acc.push(containerItem);
+
+    return acc;
+  }, []);
+}
+
+export default {
+  readJsonForm(config) {
```
Comment (102) [inline frontend/services/config_reader.js]: Shouldn't we rename this functions as well? `readJsonForm` and `readJsonModal` aren't just for reading, they're for formatting

File: frontend/services/config_reader.js [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+  Object.keys(finalForm)
+    .filter(key => finalForm[key].items && !Array.isArray(finalForm[key].items))
+    .forEach(key => {
+      finalForm[key].items = map(finalForm[key].items, (item, name) => Object.assign({ name }, item));
+    });
+
+  return finalForm;
+}
+
+function configureFormItemsBySchema(items, config) {
```
Comment (102) [inline frontend/services/config_reader.js]: I also think that `formatFormFieldsItemsToArray` and `configureFormItemsBySchema` functions should contain some comments about what are they doing and what are they expecting to see as arguments (neither function name nor flow are really straightforward about this)

File: public/examples/forms/modal.json [hunk 15]
```diff
@@ -3,352 +3,375 @@
...
+          "name1": {
+            "label": "Содержание жилья",
+            "value": "0.00"
+          },
+          "name2": {
+            "label": "Пеня",
+            "value": "0.00"
+          },
+          "name3": {
+            "label": "Антена",
```
Comment (mvahrameev) [inline public/examples/forms/modal.json]: `Антенна`

File: frontend/services/config_reader.js [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+    finalConfig.form = formatFormFieldsItemsToArray(finalConfig.form);
+    finalConfig.items = configureFormItemsBySchema(finalConfig.schema, finalConfig);
+
+    return finalConfig;
+  },
+
+  createModalConfig(config) {
+    const finalConfig = cloneDeep(config);
+    const bodyJsonWithFormatedForm = finalConfig.body;
+    const footerJsonWithFormatedForm = finalConfig.footer;
```
Comment (102) [inline frontend/services/config_reader.js]: `formatted`, not `formated`

**Pull request merged**